parser/Builder.cc

```diff
@@ -1271,6 +1448,52 @@ class Builder::Impl {
         }
     }
 
+    void checkDuplicatePatternVariable(std::string name, core::LocOffsets loc) {
+        ENFORCE(name.size() > 0, "Empty pattern variable name");
+        if (name[0] == '_') {
+            return;
+        }
+
+        if (driver_->pattern_variables.declared(name)) {
+            error(ruby_parser::dclass::PatternDuplicateVariable, loc, name);
+        }
+
+        driver_->pattern_variables.declare(name);
+    }
+
+    void checkDuplicatePatternKey(std::string name, core::LocOffsets loc) {
+        if (driver_->pattern_hash_keys.declared(name)) {
+            error(ruby_parser::dclass::PatternDuplicateKey, loc, name);
+        }
+
+        driver_->pattern_hash_keys.declare(name);
+    }
+
+    void checkLVarName(std::string name, core::LocOffsets loc) {
+        std::regex lvar_regex("^[a-z_][a-zA-Z0-9_]*$");
+        if (!std::regex_match(name, lvar_regex)) {
+            error(ruby_parser::dclass::PatternLVarName, loc, name);
+        }
```

Comment:
ooc, why is this not handled by the grammar?

Reply:
Not sure, I would say to avoid conflicts?